core/src/typecheck/mod.rs

```diff
@@ -2129,46 +2088,57 @@ fn check_field<L: Linearizer>(
 ) -> Result<(), TypecheckError> {
     linearizer.add_field_metadata(lin, field);
 
-    check_with_annot(
-        state,
-        ctxt,
-        lin,
-        linearizer,
-        &field.metadata.annotation,
-        field.value.as_ref(),
-        ty,
-        field.value.as_ref().map(|v| v.pos).unwrap_or(id.pos),
-    )
+    // If there's no annotation, we simply check the underlying value, if any.
+    if field.metadata.annotation.is_empty() {
+        if let Some(value) = field.value.as_ref() {
+            check(state, ctxt, lin, linearizer, value, ty)
+        } else {
+            // It might make sense to accept any type for a value without definition (which would
+            // act a bit like a function parameter). But for now, we play safe and implement a more
+            // restrictive rule, which is that a value without a definition has type `Dyn`
```

Comment:
Is this a change in behaviour or was this just implicit in the previous typechecking code?

Reply:
This is a change in behavior, but I think the previous behavior was accidental, and neither officially supported nor documented. The previous behavior was indeed to let the type of fields without values to be unconstrained.